### DISCUSSION TITLE
Tips for building software with generative AI

### DISCUSSION THREAD

**Discussion by mikethemelanson** [community/community#56349] [Copilot Conversations]
### Select Topic Area

General

### Body

![The ReadME Project logo](https://user-images.githubusercontent.com/8250055/101071884-0efd3e00-3552-11eb-8523-11305d3cc848.png)
***
When I first started chatting with Anton Mirhorodchenko (@GreenWizard2015) earlier this year for a [featured article](https://github.com/readme/featured/open-source-accessibility?src=rmpcod) I was working on, he explained how it was that we were emailing back and forth.

>“Just imagine: I'm responding to you in Ukrainian and omitting about 60% of the letters from the words, but ChatGPT understands what I am saying and provides an answer in English.” 

He offered an example of how he composed a sentence using ChatGPT for text expansion:

>**Mirhorodchenko:** h m n s Antn, m f ukr. im dev w cplsy. 
>**ChatGPT:** I think you are trying to say: "Hi, my name is Anton. I'm from Ukraine. I'm a developer with cerebral palsy. `

Anton has difficulty typing because he has cerebral palsy, so he’s been exploring how to combine ChatGPT and GitHub Copilot to build software from the ground up: from setting up the environment and designing the architecture to writing the code itself. 

The community was so interested in Anton’s use that we asked him to contribute a Guide sharing more.

You can read Anton’s Guide, [*Harness the power of generative AI for software development*](https://github.com/readme/guides/coding-generative-ai?src=rmpcod), (authored with the help of ChatGPT) in The ReadME Project’s May edition. He writes about his efforts to build software using large-language models (LLMs), and offers his tips for getting the most out of tools like ChatGPT and GitHub Copilot.

>I recently conducted research on the effective use of AI for creating quality code, which I published in an open source project called [User Country Analytics](https://github.com/GreenWizard2015/User-Country-Analytics/). Instead of focusing solely on the ability of AI to solve problems, I wanted to determine AI’s ability to use best practices, including testing and other aspects. [...] Just like in real life, I started the project by consulting with an expert: a [software architect](https://github.com/GreenWizard2015/User-Country-Analytics/blob/main/AI%20Backlog/00%20-%20Software%20Architect%20Expert.md). In this case, however, the expert was not human, but rather ChatGPT playing the role of software architect.

I highly recommend checking it out. Working with @GreenWizard2015 has definitely changed the ways I've looked at using LLMs for so many things. 

❓Also, I'm curious - what tips would you add to the list? Do you have a favorite model? Are you bringing LLMs into your developer workflow somehow, whether GitHub Copilot or a custom open source project? Or maybe you combine GitHub Copilot with Codespaces, like @noahgift does for [running MLOps in Rust](https://github.com/readme/guides/rust-mlops?src=rmpcod)? Or how about accelerating your test-driven development (TDD) workflow by [using GitHub Copilot to write your tests](https://github.com/readme/guides/github-copilot-automattic?src=rmpcod), like @mokagio?

I’d love to hear what y’all are working on! It’s an exciting time for AI.

***
**The ReadME Project** is a GitHub platform dedicated to highlighting the best from the open source software community—the people and tech behind projects you use every day. To get the latest stories, articles, and best-practices from The ReadME Project delivered directly to your inbox, [subscribe to The ReadME Project newsletter](https://github.com/readme/#newsletter?src=rmpcod).

**Comment by Kikyo-chan** [comment]
![139121_large (1)](https://github.com/community/community/assets/18164716/75aee339-afd4-43f6-bbf4-171dce1a0334)

给你一个小心心

**Comment by Ahmadraza889** [comment]
Very good

**Comment by crguezl** [comment]
> I'm curious - what tips would you add to the list?

@mikethemelanson   

In my first interactions with the AI I used to analyze its response and modify it to make it correct. 
Over time I have realized that if you **need other people  to reproduce the chat**, this methodology is not the most appropriate.
It is sometimes better to reformulate/refine the questionnaire so that the probability that the AI generates a correct answer increases. 

Thus my proposal is: **Do the refinement, refactoring, debugging and optimization processes not so much on the initially AI-generated code, but on the question script**.

As a consequence of this:

* If you want the questionnaire to be reproducible , save the list of questions  on a markdown file
* Use markdown and $\LaTeX{}$ for communication with the AI. It has lot of advantages: you are stating what is code and what is a math formula, what you emphasizes etc. Copilot and Chat GPT answer seems to be influenced by the markdown notation.
* Refine the list of questions until they lead Copilot to a correct solution with high probability
* **"the program" is the question script** not the IA generated code

**Reply by mr-sagarnath** [reply]
I'm curious...

**Comment by tphundey** [comment]
good d

**Comment by anni-anni** [comment]
**Generative AI** is improving at such a fast pace that I doubt whether after sometime, we would simply have to put in proper prompts and the AI itself will be able to code another AI. Just some days back I used the Bard AI to code some simple lines on how I could integrate my external [AI dubbing](https://wavel.ai/solutions/dubbing/) / text-to-speech software with my smartphone at home that my family uses. I used teh codes and didn't work as well as I wanted it too, but nonetheless, it was impressive 👍🏽

**Comment by 100Khushi** [comment]
Thanks for sharing Anton's impactful use of ChatGPT in software development! It's amazing to see how AI techniques like [voiceover](https://wavel.ai/solutions/voiceover) or text to speech is aiding individuals like Anton in overcoming unique challenges.

❓ How are you incorporating large-language models (LLMs) such as ChatGPT or GitHub Copilot into your development workflow? Share your favorite models and tips! Whether it's MLOps, custom projects, or enhancing test-driven development (TDD), let's chat about the exciting intersection of human creativity and AI in coding. Explore more about AI voice over in development.

**Comment by james9832** [comment]
When building software with generative AI, define clear objectives, use high-quality training data, ensure ethical guidelines, and prioritize user safety [here](https://cmp-punjab.com.pk/bisp-8171-check-online-cnic-payment-eligibility/#google_vignette). Continuously test outputs, monitor performance, and integrate human feedback. Scalability, transparency, and responsible deployment are key for long-term success.

**Comment by ali4353035-alt** [comment]
Anton’s story is truly inspiring and shows how AI tools like ChatGPT and GitHub Copilot can break barriers for developers, making software creation more accessible than ever. Leveraging AI for everything from architecture to testing is a game-changer for productivity and innovation. For more tech insights and resources, visit https://apksmoke.net/.

**Comment by infodumpwill-beep** [comment]
This tool is crucial for the 9 million South Africans, especially the unemployed youth under 35, who depend on the grant. The SASSA website also offers   https://checkmy-grant.co.za/  alternative ways to check the status via WhatsApp, mobile apps, and other platforms for wider accessibility.

**Comment by ta714263-glitch** [comment]
Selain itu, Delima 2.0 membolehkan pelajar menebus Baucar Buku Digital MADANI bernilai RM50 atau RM100 melalui Kota Buku Digital. Dengan antaramuka yang mesra pengguna, platform ini memudahkan pendaftaran, log masuk, dan akses kepada bahan pembelajaran dengan selamat   [Delima 2.0 Login](https://delima2-0.my/)  . Panduan langkah demi langkah juga disediakan untuk memastikan pengalaman pembelajaran yang lancar dan sejajar dengan kemajuan teknologi terkini.

**Comment by abcdefghijklmnopqurstwxyz208-ai** [comment]
Parivahan Sewa offers a variety of services for driving licenses, including applying for new licenses, renewals, and learner’s permits. Users can schedule    https://parivahanssewa.com/  appointments, track application statuses, and take online learner’s tests. The platform also provides services like obtaining duplicate licenses and updating personal details, streamlining the entire process online.RTO Vehicle InformationVehicle electronics

**Comment by pth55** [comment]
I leverage Copilot for fast prototyping for some python and js stuff, using precise prompts to generate modular code and tests, boosting my TDD workflow by ~30%. what's your tool to boost performance?

**Comment by anwar03012446339-web** [comment]
[EPDS Bihar](https://epdssbihar.com/) बिहार एक डिजिटल प्लेटफॉर्म है जिसे बिहार सरकार ने राशन कार्ड और उससे जुड़ी सेवाओं के प्रबंधन को सरल और पारदर्शी बनाने के लिए शुरू किया है। इस पोर्टल के माध्यम से नागरिक नए राशन कार्ड के लिए आवेदन कर सकते हैं, आवेदन की स्थिति चेक कर सकते हैं और राशन कार्ड लिस्ट और RCMS    https://epdssbihar.com/   रिपोर्ट डाउनलोड कर सकते हैं। यह पोर्टल सरकारी योजनाओं के तहत खाद्यान्न और आवश्यक वस्तुओं के वितरण में पारदर्शिता और दक्षता सुनिश्चित करता है, जिससे नागरिकों का समय और प्रयास बचता है। डिजिटल सेवाओं के माध्यम से बिहार सरकार नागरिकों को राशन कार्ड संबंधित सभी सेवाएं सुलभ बना रही है।

**Comment by ik5081255-cmd** [comment]
Para estar al día con tu Beca Benito Juárez, es fundamental usar el Buscador de Estatus. Esta herramienta te permite verificar fácilmente si tu beca está activa, en revisión o si hay pagos pendientes. Simplemente ingresa tu CURP (Clave Única de Registro de Población) para acceder   https://buscadorstatus.com/   a tu información personalizada sobre tu estado. Ya sea que te preocupen las próximas fechas de pago o si tu hijo/a seguirá siendo elegible para el programa de becas, el Buscador de Estatus de Becas Benito Juárez te brinda toda la información importante que necesitas en tiempo real.

**Comment by sands21** [comment]
I'd like to highlight the importance of **context** and how context engineering can significantly boost the output quality of any generative AI model being used.

**Context engineering** is about shaping not only your prompts but also the supporting information you feed to the model, such as examples, background data, prior chat history, or even “system”/instruction blocks. Carefully curating context enables:
- More relevant, on-topic, and accurate model responses
- Improved handling of multi-turn interactions or larger workflows
- Better performance when chaining or orchestrating multiple models/tools
- More consistent outputs

From my experience, spending time on **engineering the right context window** (what to include/exclude, what reference materials to attach, etc.) has a bigger impact on output quality than tweaking the core prompt itself, especially for more complex, production-grade use cases.

For the most part, you will be good having the following files in your workspace when building anything with the help of AI:
- **architecture.md** - This defines your project file structure, data flow, state management etc.
- **tasks.md** - granular step by step task breakdown for MVP development and testing.
- **design.md** - design principles/guide you want the model to follow while designing anything.

As for my favorite models right now, I’m especially impressed by:
- **GPT‑5** (incredible versatility, reasoning and instruction following)
- **Claude Sonnet 4** (overall just amazing)
- **GLM 4.5** (super cheap and actually good performance)

I use them through **Cursor** and **Kilo Code** if you're curious.

Hope this helps!

**Comment by saghirkhan381-ux** [comment]
In the Philippines, the implementation of the SIM Registration Act (Republic Act 11934) has made it mandatory for all mobile network users to register their    https://tntsimregisterr.com/  TNT SIM card and Smart SIM cards to prevent misuse, enhance security, and combat online fraud.

**Comment by udduup810-cloud** [comment]
Selain itu, [IDME](https://id-mekpm.my/) KPM juga dilengkapi dengan ciri Single Sign-On (SSO), yang memudahkan pengguna mengakses pelbagai sistem yang terintegrasi dengan IDMS tanpa perlu mendaftar berulang kali. Sistem ini memastikan pengesahan pengguna dilakukan secara luaran, mengurangkan masalah duplikasi kelayakan pengguna di seluruh sistem dalam ekosistem pendidikan Malaysia.  https://id-mekpm.my/

**Comment by s87817708-dot** [comment]
Dilancarkan pada tahun 2024, sistem AINS [AINS NILAM ](https://ainsnilam-kmp.my/)   Delima 2.0 ini adalah sebahagian daripada inisiatif ‘mekanisme jabat tangan’ yang bertujuan menghubungkan pelbagai kementerian, jabatan, dan agensi untuk menangani isu awam dengan lebih berkesan. Sistem ini dijangka dapat meningkatkan kecekapan dalam merekod bahan bacaan pelajar sambil meringankan beban tugas guru yang sebelum ini menguruskan tugas ini secara manual.

**Comment by ronadiaokaliff-art** [comment]
[APDM](https://theapdm.my/) 2025, atau Aplikasi Pangkalan Data Murid, merupakan sistem digital terkini yang dibangunkan oleh Kementerian Pendidikan Malaysia bagi    https://theapdm.my/   memudahkan pengurusan data pelajar secara dalam talian. Ia digunakan secara meluas oleh guru dan pentadbir sekolah untuk merekod kehadiran, menyimpan maklumat pelajar, dan menyelaras laporan pendidikan harian. Dengan antaramuka mesra pengguna yang telah ditambah baik, sistem ini kini menyokong ciri keselamatan moden seperti pengesahan dua faktor dan sokongan biometrik bagi mengelakkan penyalahgunaan akses.

**Comment by guchiflora36-prog** [comment]
It offers services like mutation status checks,   [MyKasih](https://my-kasih.my//)   land conversion applications, and digitally signed records, along with tools like i-Wallet and RTC Sketch for secure management. This paperless initiative benefits landowners, farmers, and government institutions by providing quick and accurate online land details

**Comment by secreatvectoriyaa-ctrl** [comment]
It offers services like mutation status checks, land  [Bhoomi Online ](https://bhoomi-rtconline.com/)  conversion applications, and digitally signed records, along with tools like i-Wallet and RTC Sketch for secure management. This paperless initiative benefits landowners, farmers, and government institutions by providing quick and accurate online land details

**Reply by guchiflora36-prog** [reply]
> It offers services like mutation status checks,[ MyKasih](https://my-kasih.my//)  conversion applications, and digitally signed records, along with tools like i-Wallet and RTC Sketch for secure management. This paperless initiative benefits landowners, farmers, and government institutions by providing quick and accurate online land details

[a comment by an automated account is not shown]

**Comment by alimakalima07-max** [comment]
With features like the mutation status and Aadhaar linking, [Mee-Bhoomi AP](https://meebhoomi.ap.gov.in/) enhances transparency and efficiency in land management. It supports multiple languages, including Telugu and English, making it accessible to a broader audience. [Meebhoomi  ](https://meebhomi.com/)

**Comment by bridjegold-blip** [comment]
Mod Skin LOL, a third-party tool introduced in 2015, offers players a straightforward way to try out custom skins in League of Legends without any complications   [Mod LOL Skin ](https://loolskin.com/)  . Since League of Legends doesn’t have a built-in method for using custom skins, Mod Skin LOL fills this gap, providing an easy and free way for players to personalize their champions.

**Comment by usmankty39-design** [comment]
E-Shikshakosh, a pioneering initiative launched by the Bihar Education Project Council, is revolutionizing the education sector in Bihar by offering a digital platform for teachers, headmasters, education officials, and schools. This one-stop portal aims to streamline school management,    https://eshikshakoosh.com/    enhance transparency, and facilitate efficient data management. Through [E-Shikshakosh](https://eshikshakoosh.com/) secure login system, educators can access real-time data, manage attendance, and engage with key educational programs.

**Comment by salishsmell-bit** [comment]
[Doraemon X](https://dorasmonx.com/) brings the beloved characters of the iconic anime series to life in an engaging Android game. Players control Nobita    https://dorasmonx.com/   and interact with characters like Shizuka and Doraemon, using gadgets like the Bamboo Copter and Anywhere Door to solve puzzles and complete missions.

**Comment by ueerawais-sudo** [comment]
[Delta Executor](https://deltaexecutoor.com/delta-executor/) , también conocido como Delta X, es una herramienta avanzada para ejecutar scripts personalizados en Roblox, mejorando la experiencia de juego con características como teletransportación y trucos de velocidad. Es compatible con Android  https://deltaexecutoor.com/  , iOS y PC, y no requiere un sistema de llaves, permitiendo un acceso instantáneo. Utiliza el motor de ejecución de scripts Aorom, compatible con el 92% de UNC, lo que minimiza el riesgo de detección.

**Comment by stunnyriver-prog** [comment]
Selamat datang ke [SemaakanBantuan](https://semaakanbantuan.my/semaakanbantuan/), laman web rujukan utama anda untuk semua maklumat berkaitan [ SemaakanBantuan](https://semaakanbantuan.my/)

**Comment by danishzamoorrd-prog** [comment]
[EPDS Bihar](https://epdsbihaar.com/) बिहार राज्य सरकार की एक डिजिटल पहल है, जो सब्सिडी वाले खाद्य और आवश्यक वस्तुओं के वितरण को सरल और पारदर्शी बनाती है। इस प्लेटफ़ॉर्म के माध्यम से नागरिक राशन कार्ड के लिए आवेदन कर सकते हैं, आवेदन की स्थिति देख सकते हैं, और राशन कार्ड की सूची डाउनलोड कर सकते हैं। https://epdsbihaar.com/

**Comment by cloudesstain-creator** [comment]
The ReadME Project by GitHub celebrates the heart of open source—its people and their ideas. It’s a home for the stories, insights, and innovations shaping the software you rely on every day. From in-depth articles to practical guides, The ReadME Project shines a light on the creators driving the future of open source.

Subscribe to get fresh perspectives, inspiring stories, and proven best practices from the global open source community—delivered straight to your inbox.  [EPDS Bihar ](https://epdsbiharr.com/)

**Comment by razorwavewistfulvibes-create** [comment]
The ReadME Project by     [UP Ration Card ](https://rationcard-download.com/)  GitHub honors the essence of open source—its people and their powerful ideas. It’s a dedicated space for the stories, lessons, and breakthroughs that define the software shaping our world. Through thoughtful features, deep dives, and hands-on guides, The ReadME Project spotlights the innovators and communities building the future of open source collaboration.

**Comment by markbrown3915-hue** [comment]
[IDME KPM](https://theidmekpm.my/) adalah platform digital berpusat Kementerian Pendidikan Malaysia yang memudahkan pengurusan identiti dan pengesahan data dalam sektor pendidikan. Sistem ini menyediakan Single Sign-On (SSO) untuk mengakses pelbagai aplikasi pendidikan seperti DELIMa dan eRPH. Ibu bapa boleh mendaftarkan anak dan mengurus penempatan sekolah, guru boleh mengurus pengajaran dan rekod pelajar, manakala pentadbir sekolah boleh menyelia maklumat sekolah, guru, dan pelajar. IDME KPM memastikan keselamatan data dan pengurusan pendidikan yang cekap dalam ekosistem digital Malaysia.

**Comment by abhiyazeeshan-del** [comment]
मध्यप्रदेश सरकार द्वारा शुरू किया गया समग्र पोर्टल नागरिकों को एक अद्वितीय [Samagra ID](https://samagra-idportal.info/) प्रदान करता है, जिससे विभिन्न कल्याण योजनाओं और सेवाओं तक पहुंच सरल   https://samagra-idportal.info/

**Comment by maheenferoaz-afk** [comment]
[DELIMa 2.0](https://iddeelima.my/delima-2-0/) , inisiatif oleh Kementerian Pendidikan Malaysia, adalah platform pembelajaran dalam talian yang direka untuk mengintegrasikan teknologi  [Delima 2.0 KPM id login](https://iddeelima.my/)

**Reply by hackjawal** [reply]
[شات عراقنا](https://iraqna.chat/)

**Comment by najamqariat-oss** [comment]
[Sewayojan Portal](https://sewayojan-portal.com/) : Registration, Login, Online Apply, Job Search उत्तर प्रदेश सरकार द्वारा शुरू किया गया एक डिजिटल प्लेटफ़ॉर्म है, जिसे sewayojan.up.nic.in  https://sewayojan-portal.com/

**Comment by bakhtawarshamss073-commits** [comment]
[MPBhulekh](https://mp-bhulekh.info/) पोर्टल को मध्यप्रदेश सरकार ने नागरिकों के लिए भूमि रिकॉर्ड्स की प्राप्ति को सरल और पारदर्शी बनाने के लिए शुरू किया है। यह पोर्टल अब नागरिकों को खसरा, खतौनी, B1 फॉर्म और भू नक्शा (Bhu Naksha) जैसी महत्वपूर्ण भूमि जानकारी आसानी से ऑनलाइन उपलब्ध कराता है। यह कदम भूमि संबंधित सेवाओं में सुधार और सरकारी कार्यालयों के चक्कर काटने की आवश्यकता को कम करने के लिए उठाया गया है।    https://mp-bhulekh.info/

**Comment by alaiyaasad-cmyk** [comment]
ट्रांसफर व प्रमोशन के साथ-साथ सरकारी शिक्षा योजनाओं और सांख्यिकीय रिपोर्टों के अपडेट शामिल हैं। पारदर्शिता और जवाबदेही को ध्यान में रखकर बनाया गया यह पोर्टल अभिभावकों, अधिकारि    https://shaladarpan-portal.info/   यों और नागरिकों को उपयोगकर्ता-मित्र और मोबाइल-फ्रेंडली इंटरफेस के जरिये कभी भी, कहीं से भी जानकारी तक पहुँचने की सुविधा देता है, जिससे अलग-अलग पोर्टल या कार्यालय जाने की आवश्यकता समाप्त हो जाती है।

**Comment by hijabw17-stack** [comment]
डिएक्टिवेट करने से पहले नंबर की सही जांच करें, क्योंकि अगर आपने गलती से अपने व्यक्तिगत SIM को डिएक्टिवेट कर दिया, तो आपको समस्या हो सकती है। TAFCOP telecom fraud को रोकने में      https://tafcoopportal.com/    मदद करता है, जिससे आपकी पहचान सुरक्षित रहती है।

**Comment by pluginfichiers-cyber** [comment]
न्यूनतम मानव हस्तक्षेप (Minimal human      https://service-plusbihar.com/  intervention) से delays या errors की संभावना घटती है, जो manual handling के कारण हो सकती हैं।

**Comment by manueekhazran-blip** [comment]
Lightroom has become a favorite among both amateurs and professionals due to its intuitive interface and powerful capabilities, making it the go-to    https://lightrm.pro/   app for creating high-quality visuals. The MOD version allows access to premium features without a subscription, offering a comprehensive editing experience.

**Comment by shahnawazzmunaza-cyber** [comment]
Beyond service access, Seva Sindhu promotes good governance by improving financial transparency, accountability, and public awareness. It ensures timely delivery of social welfare schemes and    https://sevaasindhukarnataka.com/   integrates seamlessly with the Service Plus framework, making Karnataka’s public services more efficient and inclusive.

**Comment by ayeshaakamran868-collab** [comment]
[ e Shram Card ](https://eshramcaard.com/) योजनाओं का सीधे लाभ देने का उद्देश्य रखती है, जिससे वित्तीय समावेशन और सुरक्षा को बढ़ावा मिलता है।

**Comment by ayeshaakamran868-collab** [comment]
योग्य श्रमिक ई-श्रम पोर्टल के माध्यम से ऑनलाइन पंजीकरण कर सकते हैं। एक बार पंजीकरण होने के बाद, उन्हें ₹3,000 तक की मासिक पेंशन, ₹2 लाख का मृत्यु बीमा और अन्य लाभ मिलते      https://eshramcaard.com/   हैं। यह पहल लाखों श्रमिकों को सरकारी कल्याण योजनाओं का सीधे लाभ देने का उद्देश्य रखती है, जिससे वित्तीय समावेशन और सुरक्षा को बढ़ावा मिलता है।

**Comment by jodhkrazlaiyashhh-cmd** [comment]
Recently partnering with 101Game, Daman Games has expanded its features, providing even more exciting ways to win. The platform ensures a safe and sec  https://damaanclub.com/

**Comment by rosedane810-ops** [comment]
This is such an inspiring story! Anton’s use of ChatGPT to overcome typing challenges and build software is a brilliant example of how AI can empower developers. I especially love how he combines LLMs with tools like GitHub Copilot to streamline the entire workflow—from architecture to test-driven development. I’ve been experimenting with similar setups, and it’s amazing how much time it can save once you integrate it properly. Can’t wait to [login ](https://samagraidgyaan.com/spr-login/)and check out Anton’s full Guide in The ReadME Project—definitely adding it to my reading list!

**Comment by ramnarayanank** [comment]
I like this discussion because it shows there's no single "right" way to approach development. AI has made it easier to prototype ideas, automate repetitive tasks and improve productivity, but the foundation still comes down to good architecture and clean code. 

If your goal is to [build AI software](https://www.contus.com/blog/how-to-build-ai-software/), start with a real problem to solve, validate your idea early and iterate based on user feedback. AI is a great accelerator, but thoughtful engineering is what turns a prototype into a successful product.
